lib/Target/CBackend/CBackend.cpp

```diff
@@ -5297,7 +5299,7 @@ void CWriter::printGEPExpression(Value *Ptr, gep_type_iterator I,
                               // but we don't support it here
     if (I.isStruct()) {
       Out << ".field" << cast<ConstantInt>(I.getOperand())->getZExtValue();
-    } else if (IntoT->isArrayTy()) {
+    } else if (IntoT->isArrayTy() && I != FirstI) {
```

Comment:
One of the confusing things about this code is the distinction between`I` and `IntoT` (`I.getIndexedType()`). Based on reading the [the LLVM documentation and corresponding source code](https://llvm.org/doxygen/classllvm_1_1generic__gep__type__iterator.html#a48cf1b8139cae1d146beb7e0f73c0817), I think the latter might be one level deeper than the former? But if you change the code above from `if (I.isStruct()) {` to `if (IntoT->isStructTy()) {`, all the tests still pass. In any case, the way we're mixing and matching these makes me nervous.

As for this case, I can probably change `} else if (IntoT->isArrayTy() && I != FirstI) {` to `} else if (I.isBoundedSequential() && IntoT->isArrayTy()) {`. What “bounded sequential” means is that it's an array or vector type, not the imaginary unbounded array that the first GEP index is dealing with.

Reply:
> I think [`IntoT`] might be one level deeper than [`I`]?

Actually, I think I was misunderstanding the internals of the GEP type iterator. It mystifies me why it's implemented the way it is, but it seems these refer to the same type.

Reply:
I think the code for operand zero should be adjusted slightly to always handle the index and print some index:
```julia
  Value *FirstOp = I.getOperand();
  Type *IntoT = I.getIndexedType();
  ++I;
  if (!isa<Constant>(FirstOp) || !cast<Constant>(FirstOp)->isNullValue()) {
    // First index isn't simple, print it the hard way.
    writeOperand(Ptr);
    Out << '[';
    writeOperandWithCast(FirstOp, Instruction::GetElementPtr);
    Out << ']';
  } else if (isAddressExposed(Ptr)) {
...
```

Reply:
Oh, that's a nice idea. Given the first index is special, using the same code for it as for the other indices just creates a mess, indeed.

I'll try this.

Reply:
I reworked it. I may have gotten a bit carried away with changing the comments, so if you'd prefer I simplify the commit, I understand 😛